File: python/cortex_agent/llm.py
"""
Cortex Agent LLM Provider — 多 Provider 支持
══════════════════════════════════════════════

支持 DeepSeek / OpenAI / GLM（智谱），流式 + 非流式调用。
模型别名解析 + base_url 配置 + provider 感知的 thinking 参数。
"""

import json, time, httpx
from openai import OpenAI
from typing import List, Dict, Callable, Optional, Tuple


# ══════════════════════════════════════════════════════════════
# LLM Provider
# ══════════════════════════════════════════════════════════════

class LLMProvider:
    # 默认提供者注册表：settings.json 中 providers 段可覆盖
    DEFAULT_PROVIDERS = {
        "deepseek": {
            "name": "deepseek",
            "base_url": "https://api.deepseek.com/v1",
            "models": {
                "flash": "deepseek-v4-flash",
                "pro":   "deepseek-v4-pro",
            },
        },
        "openai": {
            "name": "openai",
            "base_url": "https://api.openai.com/v1",
            "models": {},
        },
        "glm": {
            "name": "glm",
            "base_url": "https://open.bigmodel.cn/api/paas/v4",
            "models": {},  # 无别名 — glm-5.2 / glm-5.2-flash 直接作为模型名使用
        },
    }

    # ── Model Capabilities Registry ──
    # 每个模型的实际上下文窗口和最大输出 token 数。
    # context_limit=0 或 max_tokens=0 时自动从此注册表解析。
    MODEL_CAPABILITIES = {
        # DeepSeek V4 系列 — 1M 上下文窗口
        "deepseek-v4-flash":  {"context_window": 1_000_000, "max_output_tokens": 8192},
        "deepseek-v4-pro":   {"context_window": 1_000_000, "max_output_tokens": 8192},
        # GLM 系列 — 1M 上下文窗口
        "glm-5.2":            {"context_window": 1_000_000, "max_output_tokens": 4096},
        "glm-5.2-flash":     {"context_window": 1_000_000, "max_output_tokens": 4096},
        # OpenAI 系列
        "gpt-4o":             {"context_window": 128_000,   "max_output_tokens": 16384},
        "gpt-4o-mini":        {"context_window": 128_000,   "max_output_tokens": 16384},
    }
    # 按前缀匹配的默认能力（用于未注册的模型名）
    _PREFIX_DEFAULTS = [
        ("deepseek",  {"context_window": 1_000_000, "max_output_tokens": 8192}),
        ("glm",       {"context_window": 1_000_000, "max_output_tokens": 4096}),
        ("gpt-4",     {"context_window": 128_000,   "max_output_tokens": 16384}),
        ("gpt-3.5",   {"context_window": 16_000,    "max_output_tokens": 4096}),
    ]
    # 全局回退默认值
    _FALLBACK_CAPS = {"context_window": 128_000, "max_output_tokens": 8192}
    # 当前激活的提供者（首次使用前由 AgentConfig.setup_providers 初始化）
    _active = None
    _provider_name = None

    @classmethod
    def setup(cls, providers: dict = None, active: str = None):
        """从 settings.json 注册 provider 列表，并设置当前使用的 provider。
        providers 结构: { name: { base_url, api_key?, models: { alias: model_id } } }
        """
        if providers:
            for name, cfg in providers.items():
                cls.DEFAULT_PROVIDERS[name] = {
                    "name": name,
                    "base_url": cfg.get("base_url", ""),
                    "models": dict(cfg.get("models", {})),
                }
        cls._provider_name = active or "deepseek"
        cls._active = cls.DEFAULT_PROVIDERS.get(cls._provider_name, cls.DEFAULT_PROVIDERS["deepseek"])

    @classmethod
    def resolve(cls, name: str) -> str:
        """将别名解析为真实 model id。先查当前 provider 的 models 映射，再查全局。"""
        active = cls._active or cls.DEFAULT_PROVIDERS.get("deepseek", {})
        models = active.get("models", {})
        if name in models:
            return models[name]
        # 回退：跨 provider 查找
        for p in cls.DEFAULT_PROVIDERS.values():
            if name in p.get("models", {}):
                return p["models"][name]
        return name  # 按原始值传递（可能是完整 model id）

    @classmethod
    def base_url(cls) -> str:
        active = cls._active or cls.DEFAULT_PROVIDERS.get("deepseek", {})
        return active.get("base_url", "https://api.deepseek.com/v1")

    @classmethod
    def provider_name(cls) -> str:
        return cls._provider_name or "deepseek"

    @classmethod
    def resolve_capabilities(cls, model: str) -> dict:
        """解析模型的上下文窗口和最大输出 token 数。
        
        查找顺序:
          1. MODEL_CAPABILITIES 精确匹配
          2. _PREFIX_DEFAULTS 前缀匹配
          3. _FALLBACK_CAPS 回退默认值
        
        返回: {"context_window": int, "max_output_tokens": int}
        """
        model_lower = model.lower().strip()
        # 1. 精确匹配
        if model_lower in cls.MODEL_CAPABILITIES:
            return cls.MODEL_CAPABILITIES[model_lower]
        # 2. 前缀匹配
        for prefix, caps in cls._PREFIX_DEFAULTS:
            if model_lower.startswith(prefix):
                return caps
        # 3. 回退
        return cls._FALLBACK_CAPS

    def __init__(self, api_key: str, model: str, tools: List[Dict],
                 timeout: float = 60.0, max_tokens: int = 8192):
        self.client = OpenAI(api_key=api_key, base_url=self.base_url(),
                             timeout=httpx.Timeout(timeout, connect=10.0))
        self.model = model; self.tools = tools
        self.max_tokens = max_tokens
        # ── 缓存命中率追踪 ──
        self._call_count: int = 0
        self._cache_hits: int = 0
        self._total_input_tokens: int = 0
        self._total_cached_tokens: int = 0

    @property
    def cache_stats(self) -> dict:
        """返回缓存统计信息。"""
        rate = (self._cache_hits / self._call_count * 100) if self._call_count > 0 else 0
        return {
            "calls": self._call_count,
            "cache_hits": self._cache_hits,
            "hit_rate": rate,
            "total_input_tokens": self._total_input_tokens,
            "total_cached_tokens": self._total_cached_tokens,
        }

    def _track_usage(self, resp):
        """从 API 响应中提取 usage 信息更新缓存统计。"""
        try:
            usage = getattr(resp, 'usage', None)
            if usage:
                self._call_count += 1
                self._total_input_tokens += getattr(usage, 'prompt_tokens', 0) or 0
                cached = getattr(usage, 'prompt_tokens_details', None)
                if cached:
                    ct = getattr(cached, 'cached_tokens', 0) or 0
                    self._total_cached_tokens += ct
                    if ct > 0:
                        self._cache_hits += 1
        except Exception:
            self._call_count += 1  # 至少计数

    def switch(self, alias: str): self.model = self.resolve(alias)

    def _thinking_kwargs(self, thinking: bool) -> dict:
        """根据当前 provider 生成 thinking 参数。
        
        GLM 使用 thinking_budget 控制推理强度，DeepSeek/OpenAI 使用 reasoning_effort。
        """
        if not thinking:
            return {}
        provider = self.provider_name()
        if provider == "glm":
            # GLM-5.2: thinking.type + thinking_budget
            return {"extra_body": {"thinking": {"type": "enabled", "thinking_budget": "max"}}}
        # DeepSeek / OpenAI / default
        return {"extra_body": {"thinking": {"type": "enabled"}}, "reasoning_effort": "max"}

    def call(self, messages: List[Dict], thinking: bool = True
             ) -> Tuple[str, Optional[List[Dict]], Optional[str], str]:
        """非流式调用。返回 (text, tool_calls, reasoning, finish_reason)。
        
        thinking=False 时关闭推理模式，用于空响应恢复——确保 LLM 将全部
        max_tokens 预算用于 content/tool_calls 而非 reasoning。
        """
        kwargs: Dict = {
            "model": self.model, "messages": messages,
            "tools": self.tools, "max_tokens": self.max_tokens,
        }
        kwargs.update(self._thinking_kwargs(thinking))
        resp = self.client.chat.completions.create(**kwargs)
        self._track_usage(resp)
        choice = resp.choices[0]
        msg = choice.message
        text = msg.content or ""
        reasoning = getattr(msg, 'reasoning_content', None) or None
        finish_reason = choice.finish_reason or ""
        tcs = None
        if msg.tool_calls:
            tcs = [{"id": tc.id, "name": tc.function.name,
                    "args": json.loads(tc.function.arguments) if tc.function.arguments else {}}
                   for tc in msg.tool_calls]
        return text, tcs, reasoning, finish_reason

    def call_stream(self, messages: List[Dict],
                    on_text: Callable[[str], None] = None,
                    on_answer: Callable[[str], None] = None,
                    on_tool: Callable[[str, dict], None] = None,
                    thinking: bool = True
                    ) -> Tuple[str, Optional[List[Dict]], str, str]:
        """流式调用：返回 (text, tool_calls, reasoning_text, finish_reason)。
        
        thinking=False 时关闭推理模式，用于空响应恢复。
        """
        kwargs: Dict = {
            "model": self.model, "messages": messages,
            "tools": self.tools, "max_tokens": self.max_tokens,
            "stream": True,
        }
        kwargs.update(self._thinking_kwargs(thinking))
        resp = self.client.chat.completions.create(**kwargs)
        reasoning_parts, text_parts = [], []
        tool_buf: Dict[int, dict] = {}
        reasoning_done = False
        tool_seen = False
        finish_reason = ""
        for chunk in resp:
            choice = chunk.choices[0] if chunk.choices else None
            if choice and choice.finish_reason:
                finish_reason = choice.finish_reason
            delta = choice.delta if choice else None
            if not delta: continue
            if getattr(delta, 'reasoning_content', None):
                reasoning_parts.append(delta.reasoning_content)
                if on_text: on_text(delta.reasoning_content)        # thinking tokens → deep grey
            if delta.content:
                if on_answer and not reasoning_done:
                    on_answer("")  # signal: flush reasoning, switch to bright
                    reasoning_done = True
                text_parts.append(delta.content)
                if on_answer: on_answer(delta.content)               # answer tokens → bright
            if delta.tool_calls:
                if not tool_seen and on_tool:
                    on_tool("", {})  # sentinel: close reasoning before tool labels
                    tool_seen = True
                for tcd in delta.tool_calls:
                    idx = tcd.index
                    if idx not in tool_buf:
                        tool_buf[idx] = {"id": tcd.id or "", "name": "", "args_json": ""}
                    if tcd.id: tool_buf[idx]["id"] = tcd.id
                    if tcd.function:
                        if tcd.function.name: tool_buf[idx]["name"] += tcd.function.name
                        if tcd.function.arguments: tool_buf[idx]["args_json"] += tcd.function.arguments
        text = "".join(text_parts); reasoning = "".join(reasoning_parts)
        tcs = None
        if tool_buf:
            tcs = []
            for idx in sorted(tool_buf.keys()):
                tb = tool_buf[idx]
                try: args = json.loads(tb["args_json"])
                except json.JSONDecodeError: args = {}
                tcs.append({"id": tb["id"], "name": tb["name"], "args": args})
        # 流式调用：至少计数 + 估算 token
        self._call_count += 1
        total_chars = sum(len(m.get("content", "") or "") for m in messages)
        self._total_input_tokens += int(total_chars * 0.4)
        return text, tcs, reasoning, finish_reason

File: src/core/llm.ts
/**
 * LLM Provider — DeepSeek / OpenAI / GLM（智谱）流式 + 非流式
 * 与 Python llm.py 完全对应
 */
import { Message, FunctionSchema, CacheStats } from './types.js';

interface LLMConfig {
  apiKey: string;
  baseUrl: string;
  model: string;
  tools: FunctionSchema[];
  timeout: number;
  maxTokens: number;
}

interface ParsedToolCall {
  id: string;
  name: string;
  args: Record<string, unknown>;
}

export type { ParsedToolCall };

interface LLMResponse {
  text: string;
  toolCalls: ParsedToolCall[] | null;
  reasoning: string;
  finishReason: string;
}

const DEFAULT_PROVIDERS: Record<string, { baseUrl: string; models: Record<string, string> }> = {
  deepseek: {
    baseUrl: "https://api.deepseek.com/v1",
    models: { flash: "deepseek-v4-flash", pro: "deepseek-v4-pro" },
  },
  openai: {
    baseUrl: "https://api.openai.com/v1",
    models: {},
  },
  glm: {
    baseUrl: "https://open.bigmodel.cn/api/paas/v4",
    models: {},  // 无别名 — glm-5.2 / glm-5.2-flash 直接作为模型名使用
  },
};

// ── Model Capabilities Registry ──
// 每个模型的实际上下文窗口和最大输出 token 数。
// contextLimit=0 或 maxTokens=0 时自动从此注册表解析。
interface ModelCaps { contextWindow: number; maxOutputTokens: number; }

const MODEL_CAPABILITIES: Record<string, ModelCaps> = {
  // DeepSeek V4 系列 — 1M 上下文窗口
  "deepseek-v4-flash":  { contextWindow: 1_000_000, maxOutputTokens: 8192 },
  "deepseek-v4-pro":   { contextWindow: 1_000_000, maxOutputTokens: 8192 },
  // GLM 系列 — 1M 上下文窗口
  "glm-5.2":            { contextWindow: 1_000_000, maxOutputTokens: 4096 },
  "glm-5.2-flash":     { contextWindow: 1_000_000, maxOutputTokens: 4096 },
  // OpenAI 系列
  "gpt-4o":             { contextWindow: 128_000,   maxOutputTokens: 16384 },
  "gpt-4o-mini":        { contextWindow: 128_000,   maxOutputTokens: 16384 },
};

// 按前缀匹配的默认能力
const PREFIX_DEFAULTS: Array<[string, ModelCaps]> = [
  ["deepseek",  { contextWindow: 1_000_000, maxOutputTokens: 8192 }],
  ["glm",       { contextWindow: 1_000_000, maxOutputTokens: 4096 }],
  ["gpt-4",     { contextWindow: 128_000,   maxOutputTokens: 16384 }],
  ["gpt-3.5",   { contextWindow: 16_000,    maxOutputTokens: 4096 }],
];

const FALLBACK_CAPS: ModelCaps = { contextWindow: 128_000, maxOutputTokens: 8192 };

export function resolveCapabilities(model: string): ModelCaps {
  const m = model.toLowerCase().trim();
  if (MODEL_CAPABILITIES[m]) return MODEL_CAPABILITIES[m];
  for (const [prefix, caps] of PREFIX_DEFAULTS) {
    if (m.startsWith(prefix)) return caps;
  }
  return FALLBACK_CAPS;
}

let activeProvider = "deepseek";
let activeBaseUrl = "https://api.deepseek.com/v1";

export function setupProviders(providers?: Record<string, { baseUrl: string; models: Record<string, string> }>, active?: string) {
  if (providers) Object.assign(DEFAULT_PROVIDERS, providers);
  if (active) {
    activeProvider = active;
    activeBaseUrl = DEFAULT_PROVIDERS[active]?.baseUrl || activeBaseUrl;
  }
}

export function resolveModel(name: string): string {
  const provider = DEFAULT_PROVIDERS[activeProvider];
  if (provider?.models[name]) return provider.models[name];
  for (const p of Object.values(DEFAULT_PROVIDERS)) {
    if (p.models[name]) return p.models[name];
  }
  return name;
}

export class LLMProvider {
  private apiKey: string;
  private baseUrl: string;
  model: string;
  private tools: FunctionSchema[];
  private timeout: number;
  private maxTokens: number;

  /** 根据当前 provider 生成 thinking 参数。
   * GLM 使用 thinking_budget，DeepSeek/OpenAI 使用 reasoning_effort。
   */
  private _thinkingBody(thinking: boolean): Record<string, unknown> {
    if (!thinking) return {};
    if (activeProvider === "glm") {
      return { extra_body: { thinking: { type: "enabled", thinking_budget: "max" } } };
    }
    return { extra_body: { thinking: { type: "enabled" } }, reasoning_effort: "max" };
  }

  // 缓存统计
  private callCount = 0;
  private cacheHits = 0;
  private totalInputTokens = 0;
  private totalCachedTokens = 0;

  constructor(config: LLMConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.baseUrl;
    this.model = config.model;
    this.tools = config.tools;
    this.timeout = config.timeout;
    this.maxTokens = config.maxTokens;
  }

  get cacheStats(): CacheStats {
    return {
      calls: this.callCount,
      cacheHits: this.cacheHits,
      hitRate: this.callCount > 0 ? (this.cacheHits / this.callCount) * 100 : 0,
      totalInputTokens: this.totalInputTokens,
      totalCachedTokens: this.totalCachedTokens,
    };
  }

  async call(messages: Message[], thinking = true): Promise<LLMResponse> {
    const body: Record<string, unknown> = {
      model: this.model,
      messages,
      tools: this.tools.length > 0 ? this.tools : undefined,
      max_tokens: this.maxTokens,
    };
    Object.assign(body, this._thinkingBody(thinking));

    const resp = await fetch(`${this.baseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(this.timeout * 1000),
    });

    const data: Record<string, unknown> = await resp.json() as Record<string, unknown>;
    this.callCount++;
    if (data.usage) {
      const usage = data.usage as Record<string, unknown>;
      this.totalInputTokens += (usage.prompt_tokens as number) || 0;
      const details = usage.prompt_tokens_details as Record<string, unknown> | undefined;
      const cached = (details?.cached_tokens as number) || 0;
      this.totalCachedTokens += cached;
      if (cached > 0) this.cacheHits++;
    }

    const choice = (data.choices as Array<{ message?: Record<string, unknown>; finish_reason?: string }>)?.[0];
    const msg = choice?.message;
    const text: string = (msg?.content as string) || "";
    const reasoning: string = (msg?.reasoning_content as string) || "";
    const finishReason: string = choice?.finish_reason || "";

    let toolCalls: ParsedToolCall[] | null = null;
    if (msg?.tool_calls) {
      toolCalls = (msg.tool_calls as Array<{ id: string; function: { name: string; arguments: string } }>).map(tc => {
        let args: Record<string, unknown> = {};
        try { args = JSON.parse(tc.function.arguments || "{}"); } catch { /* malformed JSON from LLM */ }
        return { id: tc.id, name: tc.function.name, args };
      });
    }

    return { text, toolCalls, reasoning, finishReason };
  }

  async callStream(
    messages: Message[],
    onText?: (t: string) => void,
    onAnswer?: (t: string) => void,
    thinking = true,
  ): Promise<LLMResponse> {
    const body: Record<string, unknown> = {
      model: this.model,
      messages,
      tools: this.tools.length > 0 ? this.tools : undefined,
      max_tokens: this.maxTokens,
      stream: true,
    };
    Object.assign(body, this._thinkingBody(thinking));

    const resp = await fetch(`${this.baseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(this.timeout * 1000),
    });

    const reader = resp.body?.getReader();
    const decoder = new TextDecoder();
    const textParts: string[] = [];
    const reasoningParts: string[] = [];
    const toolBuf: Map<number, { id: string; name: string; argsJson: string }> = new Map();
    let reasoningDone = false;
    let finishReason = "";

    if (!reader) return { text: "", toolCalls: null, reasoning: "", finishReason };

    let lineBuf = "";  // 缓冲跨 chunk 的不完整 SSE 行
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      const chunk = decoder.decode(value, { stream: true });
      lineBuf += chunk;
      const lines = lineBuf.split("\n");
      // 最后一个元素可能是不完整的行，保留到下次处理
      lineBuf = lines.pop() || "";
      for (const line of lines) {
        if (!line.startsWith("data: ")) continue;
        const data = line.slice(6);
        if (data === "[DONE]") continue;
        try {
          const parsed = JSON.parse(data);
          const choice = parsed.choices?.[0];
          if (choice?.finish_reason) finishReason = choice.finish_reason;
          const delta = choice?.delta;
          if (!delta) continue;
          if (delta.reasoning_content) {
            reasoningParts.push(delta.reasoning_content);
            if (onText) onText(delta.reasoning_content);
          }
          if (delta.content) {
            if (!reasoningDone && onAnswer) { onAnswer(""); reasoningDone = true; }
            textParts.push(delta.content);
            if (onAnswer) onAnswer(delta.content);
          }
          if (delta.tool_calls) {
            for (const tcd of delta.tool_calls) {
              const idx = tcd.index;
              if (!toolBuf.has(idx)) toolBuf.set(idx, { id: "", name: "", argsJson: "" });
              const tb = toolBuf.get(idx)!;
              if (tcd.id) tb.id = tcd.id;
              if (tcd.function?.name) tb.name += tcd.function.name;
              if (tcd.function?.arguments) tb.argsJson += tcd.function.arguments;
            }
          }
        } catch { /* skip malformed chunks */ }
      }
    }
    // 处理缓冲区中剩余的最后一行
    if (lineBuf.startsWith("data: ")) {
      const data = lineBuf.slice(6);
      if (data !== "[DONE]") {
        try {
          const parsed = JSON.parse(data);
          const choice = parsed.choices?.[0];
          if (choice?.finish_reason) finishReason = choice.finish_reason;
          const delta = choice?.delta;
          if (delta?.reasoning_content) {
            reasoningParts.push(delta.reasoning_content);
            if (onText) onText(delta.reasoning_content);
          }
          if (delta?.content) {
            if (!reasoningDone && onAnswer) { onAnswer(""); reasoningDone = true; }
            textParts.push(delta.content);
            if (onAnswer) onAnswer(delta.content);
          }
          if (delta?.tool_calls) {
            for (const tcd of delta.tool_calls) {
              const idx = tcd.index;
              if (!toolBuf.has(idx)) toolBuf.set(idx, { id: "", name: "", argsJson: "" });
              const tb = toolBuf.get(idx)!;
              if (tcd.id) tb.id = tcd.id;
              if (tcd.function?.name) tb.name += tcd.function.name;
              if (tcd.function?.arguments) tb.argsJson += tcd.function.arguments;
            }
          }
        } catch { /* skip malformed */ }
      }
    }

    this.callCount++;
    const totalChars = messages.reduce((sum, m) => sum + (m.content?.length || 0), 0);
    this.totalInputTokens += Math.floor(totalChars * 0.4);

    const text = textParts.join("");
    const reasoning = reasoningParts.join("");
    let toolCalls: ParsedToolCall[] | null = null;
    if (toolBuf.size > 0) {
      toolCalls = [];
      for (const [, tb] of [...toolBuf.entries()].sort((a, b) => a[0] - b[0])) {
        let args: Record<string, unknown> = {};
        try { args = JSON.parse(tb.argsJson || "{}"); } catch { /* malformed JSON from LLM streaming */ }
        toolCalls.push({
          id: tb.id,
          name: tb.name,
          args,
        });
      }
    }
    return { text, toolCalls, reasoning, finishReason };
  }

  switch(alias: string): void {
    this.model = resolveModel(alias);
  }

  updateMaxTokens(maxTokens: number): void {
    this.maxTokens = maxTokens;
  }

  static resolve = resolveModel;
  static setupProviders = setupProviders;
}
